### DISCUSSION TITLE
TreeView. Animate when updating the list

### DISCUSSION THREAD

**Discussion by Frillaut** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3289] [Q&A] [answered]
Hi!
Can I turn animation off in TreeView? Not that she's bad, she's wonderful! But when the list is updated, it repeats the animation of opening the list.

**Comment by Frillaut** [comment]
Mmm?

**Comment by nicolaihenriksen** [comment]
@Frillaut Can you elaborate a bit on which animation you're referring to? The demo app contains an ability to manipulate the contents of a `TreeView`, and I don't see the animation repeating when items are added there?

![TreeViewAnimation](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/48845852-d2e0-4e9d-9205-d92ada3b2dc4)

**Reply by Frillaut** [reply]
I found the reason for the malfunction when adding. Sorting included animation. But more... If you enable `VirtualizingStackPanel.IsVirtualizing`, the list is scrolled to re-animate the open items. Any ideas?

**Comment by nicolaihenriksen** [comment]
@Frillaut I still can't repro the issue with `VirtualizingStackPanel.IsVirtualizing="True"`. Can you provide a sample that illustrates the issue you're experiencing?

![TreeViewAnimation](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/cfc8ec9a-dc39-4889-bdb0-ef618098da57)

**Reply by Frillaut** [reply]
With the addition, I figured it out! Now the other is a little bit. In the left list, when scrolling, the animation will be repeated. Everything is fine with the right

```
<Grid>
    <Grid.ColumnDefinitions>
        <ColumnDefinition />
        <ColumnDefinition />
    </Grid.ColumnDefinitions>
    <TreeView Grid.Column="0"
              ScrollViewer.CanContentScroll="True"
              VirtualizingStackPanel.IsVirtualizing="True"
              VirtualizingStackPanel.VirtualizationMode="Recycling"
              ItemsSource="{Binding ListView}">
        <TreeView.ItemTemplate>
            <HierarchicalDataTemplate ItemsSource="{Binding ListView}">
                <TextBlock Text="{Binding Text}" />
            </HierarchicalDataTemplate>
        </TreeView.ItemTemplate>
    </TreeView>
    <TreeView Grid.Column="1"
              ItemsSource="{Binding ListView}">
        <TreeView.ItemTemplate>
            <HierarchicalDataTemplate ItemsSource="{Binding ListView}">
                <TextBlock Text="{Binding Text}" />
            </HierarchicalDataTemplate>
        </TreeView.ItemTemplate>
    </TreeView>
</Grid>
```


```
public ICollectionView ListView { get; set; }
private ObservableCollection<TreeItem> _List = new();
public ObservableCollection<TreeItem> List { get => _List; set => Set(ref _List, value); }
```

```
public class TreeItem : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;
    protected virtual void OnPropertyChanged([CallerMemberName] string? PropertyName = null) => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(PropertyName));
    protected virtual bool Set<T>(ref T field, T value, [CallerMemberName] string? PropertyName = null)
    {
        if (Equals(field, value)) return false;
        field = value;
        OnPropertyChanged(PropertyName);
        return true;
    }

    private string _Text;
    public string Text { get => _Text; set => Set(ref _Text, value); }

    public ICollectionView ListView { get; set; }
    private ObservableCollection<TreeItem> _List = new();
    public ObservableCollection<TreeItem> List { get => _List; set => Set(ref _List, value); }

    public TreeItem(string text)
    {
        ListView = CollectionViewSource.GetDefaultView(List);
        Text = text;
    }
}
```

**Comment by nicolaihenriksen** [comment]
@Frillaut I have tried to use your code to create a sample where I can reproduce the issue. Using the code provided, I don't actually see anything other than the root node(s), so I had to make some modifications to make sub-nodes show. And after having done that, I still don't see any animations when dynamically adding items to the collection at runtime.

Are you able to share a bit more of your setup? The view model (or code-behind) backing the `TreeView.ItemsSource` where you populate the initial tree? And also the code you use to add an item which you say is causing the animation to occur.

**Reply by Frillaut** [reply]
This is how I added the elements.

```
for (int i = 0; i < 50; i++)
{
    TreeItem f = new("Folder");
    for (int e = 0; e < 10; e++) f.List.Add(new("Element"));
    List.Add(f);
}
ListView = CollectionViewSource.GetDefaultView(List);
```


This is how I check...

![TreeViewAnimation](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/116347513/c400ec4e-e416-4e19-989a-7bb68f08e311)


And if you open a lot of elements, then the thumb becomes just crazy...

![TreeViewAnimation2](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/116347513/7a42eb3e-5986-485b-886f-2234bbdebce7)


But, as I understand it, **virtualization** is not good friends with **treeview**

**Reply by nicolaihenriksen** [reply]
Ahh, I was completely misunderstanding what you meant then. I'll give this a shot.

And you're right, for the `TreeView`, virtualization only really works for the root-level nodes to my knowledge. So if you have a deep tree, where a lot of (non-root) nodes are expanded, scrolling becomes sluggish. That is at least my experience. The "heavier" your item template is, the worse it gets. With just a `TextBlock` the issue is not that bad.

**Reply by nicolaihenriksen** [reply]
@Frillaut I must be doing something different than you. The expanded nodes do not re-animate when when they are scrolled into view when I try your code.

Which version of MDIX are you using? I tried this on the latest codebase. Perhaps you could try one of the latest nightly builds to see if the problem reproduces with that?

![TreeViewAnimation](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/19572699/6015d627-8eb7-4f44-8f39-e056c121c1a5)

**Reply by Frillaut** [reply]
Version 4.9.0.
I will use TreeView without virtualization.
Thank you for your attention.

**Reply by Frillaut** [reply]
When there is no style, then thumb behaves well when virtualized. It seems to me that this behavior is due to animation.
As for your examples. When you move smoothly, it is not so noticeable, or not at all noticeable. Or virtualization is disabled)

**Reply by nicolaihenriksen** [reply]
I purposely slowed down my motions in the GIF to ensure it was correctly captured. However when I run it, there is no difference whether I move slow or fast. There is no re-animation when things are scrolled into view. I used the code from your sample so virtualization is at least attempted enabled:

```xaml
<Grid>
  <Grid.ColumnDefinitions>
    <ColumnDefinition />
    <ColumnDefinition />
  </Grid.ColumnDefinitions>
  <TreeView Grid.Column="0"
            ScrollViewer.CanContentScroll="True"
            VirtualizingStackPanel.IsVirtualizing="True"
            VirtualizingStackPanel.VirtualizationMode="Recycling"
            ItemsSource="{Binding ListView}">
    <TreeView.ItemTemplate>
      <HierarchicalDataTemplate ItemsSource="{Binding ListView}">
        <TextBlock Text="{Binding Text}" />
      </HierarchicalDataTemplate>
    </TreeView.ItemTemplate>
  </TreeView>
  <TreeView Grid.Column="1"
            ItemsSource="{Binding ListView}">
    <TreeView.ItemTemplate>
      <HierarchicalDataTemplate ItemsSource="{Binding ListView}">
        <TextBlock Text="{Binding Text}" />
      </HierarchicalDataTemplate>
    </TreeView.ItemTemplate>
  </TreeView>
</Grid>
```

I think for me to continue looking into this, I would need you to produce a small sample application that shows the issue. ZIP it up and attach it here (or create a repo or similar where I can access it).

**Reply by Frillaut** [reply]
[TreeViewTest.zip](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/files/12334078/TreeViewTest.zip)

**Reply by Frillaut** [reply]
And that's what I do.

![TreeViewAnimation3](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/assets/116347513/a8eba343-ae2d-4416-b9d1-2f44687c3f1d)

**Reply by Frillaut** [reply]
Mmm?

**Reply by nicolaihenriksen** [reply]
Thanks for the sample. My life has been overly busy recently so have had no freetime to look into this. I will get to it eventually though...

**Reply by Frillaut** [reply]
Thank you!

**Reply by nicolaihenriksen** [reply, marked as answer]
@Frillaut I had a look at your sample, and I can indeed reproduce the animation that occurs when `IsVirtualizing=True` and `VirtualizationMode=Recycling`. Changing to `VirtualizationMode=Standard` "fixes" the issue, just like setting `IsVirtualizing=False`. However, I assume you would like the settings applied, and avoid the animations.

Unfortunately, the [MDIX style/template](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/c12ada91708c63f7c47e80b1f9cc6dfa502b7cb0/MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.TreeView.xaml#L173-L406) for the `TreeViewItem` explicitly applies these animations using visual states and visual transitions. These do not lend themselves well to extendibility, nor do they allow bindings (so an attached property could be used). So the only real option you're left with, apart from manipulating the visual states in runtime in your code (which is probably not trivial!), is to make your own local copy of the `Style`. Since the style kind of relies on these animations being in place in order for it to function properly, I think your best option might be to simply drop the `Duration` on the relevant animations to 0. My guess would be that it is mostly the "expanding/collapsing" of the child content that is relevant to mock with, and in that case you need to tweak the animations that operate on the `ItemsHost` and `ScaleHost` targets. I have highlighted two of them [here](https://github.com/MaterialDesignInXAML/MaterialDesignInXamlToolkit/blob/c12ada91708c63f7c47e80b1f9cc6dfa502b7cb0/MaterialDesignThemes.Wpf/Themes/MaterialDesignTheme.TreeView.xaml#L242-L249), but there are actually a number of places you may need to tweak.

Depending on how you're using MDIX, you will need to inject this style into a `ResourceDictionary` somewhere in the visual tree. I believe you will need to add it as an implicit style (i.e. without `x:Key`) for it to be used; I don't believe you can reuse the "MaterialDesignTreeViewItem" key because of the way the resources are wired up.

Here is an example of adding in the resource locally within your `TreeView`:

```xaml
<TreeView ScrollViewer.CanContentScroll="True"
          VirtualizingStackPanel.IsVirtualizing="True"
          VirtualizingStackPanel.VirtualizationMode="Recycling"
          ItemsSource="{Binding ListView}">
    <TreeView.Resources>
        <Style TargetType="{x:Type TreeViewItem}">
          <!-- TODO: Replace with copy of the current style with relevant animation durations dropped to 0 -->
        </Style>
    </TreeView.Resources>
    <TreeView.ItemTemplate>
        <HierarchicalDataTemplate ItemsSource="{Binding ListView}">
            <TextBlock Text="{Binding Text}" />
        </HierarchicalDataTemplate>
    </TreeView.ItemTemplate>
</TreeView>
```

Probably not the answer you were hoping for, but I hope it guides you in the right direction for solving your problem.

**Reply by Frillaut** [reply]
Thank you for your time. This should help me!
